### PR TITLE
fix: try to scroll inline preview into view

### PR DESCRIPTION
**Pull request by akoreman**
*Issue #, if available:* NA

*Description of changes:* Currently, when inline preview is used at the bottom of the file, it will be invisible. This changes it to scroll down enough to get the inline preview into view. If the inline preview is bigger than the view, we scroll down so that the cursor is at the top of the screen to get the most possible into view

https://github.com/ajaxorg/ace/assets/34573235/9071dfcc-4127-4806-abb3-697c0ce2dcd0

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/virtual_renderer.js [hunk 1]
```diff
@@ -1620,6 +1620,24 @@ class VirtualRenderer {
                 className: "ace_ghost_text"
             };
             this.session.widgetManager.addLineWidget(this.$ghostTextWidget);
+
+            // Check wether the line widget fits in the part of the screen currently in view
+            var pixelPosition = this.$cursorLayer.getPixelPosition(insertPosition, true);
+            var el = this.container;
+            var height = el.clientHeight;
```
Comment (andredcoliveira) [inline src/virtual_renderer.js:1627]: `clientHeight` might be rounded down. Maybe using `getBoundingClientRect` would be a little more robust. Though I'm not sure how prominent fractional heights are in ace, if at all.

From https://developer.mozilla.org/en-US/docs/Web/API/Element/clientHeight

File: src/virtual_renderer.js [hunk 2]
```diff
@@ -1620,6 +1620,24 @@ class VirtualRenderer {
...
+            var fitsY = textLines.length * this.lineHeight < height - pixelPosition.top;
+
+            // If it fits, no action needed
+            if (fitsY) return;
+            
+            // If it can fully fit in the screen, scroll down until it fits on the screen
+            // if it cannot fully fit, scroll so that the cursor is at the top of the screen
+            // to fit as much as possible.
+            if (textLines.length * this.lineHeight < height) {
```
Comment (andredcoliveira) [inline src/virtual_renderer.js:1636]: **nitpick:** we multiply `textLines.length * this.lineHeight` twice, so we can store the value and reuse it across lines 1628 and 1636.

File: src/autocomplete.js [hunk 3]
```diff
@@ -160,9 +162,14 @@ class Autocomplete {
                 this.tooltipTimer.call(null, null);
                 return;
             }
+
+            // Update the popup position after a short wait to account for potential scrolling
+            this.popupTimer.schedule();
+            this.tooltipTimer.schedule();
+        } else {
+            this.$updatePopupPosition();
```
[the remark this answers is not in the document]
Comment (akoreman) [reply src/autocomplete.js:170]: `this.tooltipTimer` updates the position of the doc tooltip, `this.popupTimer` updates the position of the popup itself. They are both `lang.delayedCall`s which can be scheduled to happen after a timeout to allow some time for other stuff to happen.

**Review by andredcoliveira — APPROVED** [review]
Looks good. Left a couple of small comments.

File: src/autocomplete.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (whazor) [reply src/autocomplete.js:170]: Why does it happen outside the `this.inlineRenderer && this.inlineEnabled` case? Is it to cancel if inline rendering gets disabled? Or is it a general improvement for the non-inline case? If so, we should update the GH description.

[the remark this answers is not in the document]
Comment (akoreman) [reply src/autocomplete.js]: tried to make this a bit more clear in a new commit

File: src/virtual_renderer.js (hunk 1 above)
Comment (akoreman) [reply src/virtual_renderer.js]: changed this

File: src/virtual_renderer.js (hunk 2 above)
Comment (akoreman) [reply src/virtual_renderer.js]: changed this

**Review by andredcoliveira — APPROVED** [review]

**Pull request merged**